oauth-client/src/main/java/io/strimzi/kafka/oauth/client/JaasClientOauthLoginCallbackHandler.java

```diff
@@ -0,0 +1,162 @@
+import static io.strimzi.kafka.oauth.common.OAuthAuthenticator.loginWithRefreshToken;
+
+public class JaasClientOauthLoginCallbackHandler implements AuthenticateCallbackHandler {
+
+    private static Logger log = LoggerFactory.getLogger(JaasClientOauthLoginCallbackHandler.class);
+
+    private ClientConfig config = new ClientConfig();
+
+    private String token;
+    private String refreshToken;
+    private String clientId;
+    private String clientSecret;
+    private URI tokenEndpoint;
+    private String usernameClaim;
+
+
+    @Override
+    public void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries) {
+        if (!OAuthBearerLoginModule.OAUTHBEARER_MECHANISM.equals(saslMechanism))    {
+            throw new IllegalArgumentException(String.format("Unexpected SASL mechanism: %s", saslMechanism));
+        }
+
+        for (AppConfigurationEntry e: jaasConfigEntries) {
+            if ("KafkaClient".equals(e.getLoginModuleName()) || e.getLoginModuleName() == null) {
+                Properties p = new Properties();
+                p.putAll(e.getOptions());
+                config = new ClientConfig(p);
+            }
+        }
+
+        token = config.getValue(ClientConfig.OAUTH_ACCESS_TOKEN);
+        if (token == null) {
+            String endpoint = config.getValue(ClientConfig.OAUTH_TOKEN_ENDPOINT_URI);
+
+            if (endpoint == null) {
+                throw new RuntimeException("Access Token not specified ('oauth.access.token'). OAuth2 Token Endpoint ('oauth.token.endpoint.uri') should then be set.");
+            }
+
+            try {
+                tokenEndpoint = new URI(endpoint);
+            } catch (URISyntaxException e) {
+                throw new RuntimeException("Specified token endpoint uri is invalid: " + endpoint);
+            }
+
+            refreshToken = config.getValue(ClientConfig.OAUTH_REFRESH_TOKEN);
+
+            clientId = config.getValue(Config.OAUTH_CLIENT_ID);
+            clientSecret = config.getValue(Config.OAUTH_CLIENT_SECRET);
+
+            if (clientId == null) {
+                throw new RuntimeException("No client id specified ('oauth.client.id')");
+            }
+
+            if (refreshToken == null && clientSecret == null) {
+                throw new RuntimeException("No access token, refresh token, nor client secret specified");
+            }
+        }
+
+        usernameClaim = config.getValue(Config.OAUTH_USERNAME_CLAIM, "sub");
+        if ("sub".equals(usernameClaim)) {
+            usernameClaim = null;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Configured JaasClientOauthLoginCallbackHandler:\n    token: " + token
+                    + "\n    refreshToken: " + refreshToken + "\n    tokenEndpointUri: " + tokenEndpoint
+                    + "\n    clientId: " + clientId + "\n    clientSecret: " + clientSecret
+                    + "\n    usernameClaim: " + usernameClaim);
```

Comment:
placeholders insted of string concatanation?

Reply:
Generally that makes sense in order to skip redundant text concatenations, but I think in this case the code block is well protected behind isDebugEnabled() check and it seems to me that templating strings would actually mean more code, less readable code, and also slightly slower code.